muikku-school-data-pyramus/src/main/java/fi/otavanopisto/muikku/plugins/schooldatapyramus/PyramusUpdater.java

```diff
@@ -606,22 +606,25 @@ public boolean updateCourseStudent(Long courseStudentId, Long courseId, Long stu
       SchoolDataIdentifier workspaceUserIdentifier = new SchoolDataIdentifier(identifier, SchoolDataPyramusPluginDescriptor.SCHOOL_DATA_SOURCE);
       WorkspaceUserEntity workspaceUserEntity = workspaceUserEntityController.findWorkspaceUserEntityByWorkspaceUserIdentifierIncludeArchived(workspaceUserIdentifier);
       CourseStudent courseStudent = pyramusClient.get().get("/courses/courses/" + courseId + "/students/" + courseStudentId, CourseStudent.class);
-      if (courseStudent != null && !courseStudent.getArchived()) {
-        boolean studentActive = isStudentActive(courseStudent.getStudentId());
-        
-        if (workspaceUserEntity == null && studentActive) {
+
+      // #3161: Refactored logic for syncing course students
+      
+      if (workspaceUserEntity != null && (courseStudent == null || courseStudent.getArchived())) {
+        // If course student has been removed in Pyramus, permanently delete it from Muikku (see #3091)
+        fireCourseStudentRemoved(courseStudentId, studentId, courseId);
+      }
+      else {
+        if (workspaceUserEntity == null) {
+          // If course student exists in Pyramus but not in Muikku, create
           fireCourseStudentDiscovered(courseStudent);
-          return true;
-        } else {
-          if (studentActive) {
-            fireCourseStudentUpdated(courseStudent);
-          } else {
-            fireCourseStudentRemoved(courseStudentId, studentId, courseId);
-          }
         }
-      } else {
-        if (workspaceUserEntity != null) {
-          fireCourseStudentRemoved(courseStudentId, studentId, courseId);
+        else {
+          // Both sides exist, so update
+          fireCourseStudentUpdated(courseStudent);
+          // If the student is inactive (studies have ended), archive
+          if (!workspaceUserEntity.getArchived() && !isStudentActive(courseStudent.getStudentId())) {
+            workspaceUserEntityController.archiveWorkspaceUserEntity(workspaceUserEntity);
```

Comment:
This straight call feels out of place - the events are also used for updating search indexes so it would probably be preferred place to handle this. Is it a remove event or should the update event handle it, dunno.